genie-web/src/test/groovy/com/netflix/genie/web/services/impl/JobKillServiceImplSpec.groovy

```diff
@@ -0,0 +1,100 @@
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.services.impl
+
+import com.netflix.genie.common.dto.JobStatus
+import com.netflix.genie.common.exceptions.GenieException
+import com.netflix.genie.common.exceptions.GenieServerException
+import com.netflix.genie.common.internal.exceptions.unchecked.GenieJobNotFoundException
+import com.netflix.genie.test.categories.UnitTest
+import com.netflix.genie.web.services.JobKillServiceV4
+import com.netflix.genie.web.services.JobPersistenceService
+import com.netflix.genie.web.services.JobSearchService
+import org.junit.experimental.categories.Category
+import spock.lang.Specification
+
+/**
+ * Tests for JobKillServiceImpl
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Category(UnitTest.class)
+class JobKillServiceImplSpec extends Specification {
+    JobKillServiceV3 jobKillServiceV3
+    JobKillServiceImpl service
+    JobPersistenceService jobPersistenceService
+    JobKillServiceV4 jobKillServiceV4
+    String jobId
+
+    void setup() {
+        jobPersistenceService = Mock()
+        jobKillServiceV4 = Mock()
+        jobKillServiceV3 = Mock()
+        jobId = UUID.randomUUID().toString()
+        service = new JobKillServiceImpl (jobKillServiceV3, jobKillServiceV4, jobPersistenceService)
+    }
+
+    def "Invoke JobKillServiceV3 for a v3 job"() {
+
+        when:
+        service.killJob(jobId, "testing")
+
+        then:
+        1 * jobPersistenceService.isV4(jobId) >> false
+        0 * jobKillServiceV4.killJob(jobId, "testing")
+        1 * jobKillServiceV3.killJob(jobId, "testing")
+    }
+
+    def "Invoke JobKillServiceV4 for a v4 job"() {
+
+        when:
+        service.killJob(jobId, "testing")
+
+        then:
+        1 * jobPersistenceService.isV4(jobId) >> true
+        1 * jobKillServiceV4.killJob(jobId, "testing")
+        0 * jobKillServiceV3.killJob(jobId, "testing")
+    }
+
+    def "Percolate up the exception thrown by underlying services when killing a job"() {
+
+        when:
+        service.killJob(jobId, "testing")
+
+        then:
+        1 * jobPersistenceService.isV4(jobId) >> { throw new GenieJobNotFoundException()}
+        thrown(GenieJobNotFoundException)
+
+        when:
+        service.killJob(jobId, "testing")
+
+        then:
+        1 * jobPersistenceService.isV4(jobId) >> true
+        1 * jobKillServiceV4.killJob(jobId, "testing") >> { throw new GenieServerException("Error killing v4 job")}
+        thrown(GenieException)
+
+        when:
+        service.killJob(jobId, "testing")
+
+        then:
+        1 * jobPersistenceService.isV4(jobId) >> false
+        1 * jobKillServiceV3.killJob(jobId, "testing") >> { throw new GenieServerException("Error killing v3 job")}
+        thrown(GenieException)
```

Comment:
This could be stricter.
You could check that *exactly* the same exception, unchanged, is thrown.
Whereas you're just checking that an exception of a certain type is thrown.

Reply:
Fixed